lib/howl/interactions/external_command.moon

```diff
@@ -0,0 +1,207 @@
+
+  _updir: =>
+    if @directory.parent
+      @_chdir @directory.parent
+
+  _initialize_list_widget: =>
+    @list_widget = ListWidget nil,
+      never_shrink: true
+    @list_widget.max_height = math.floor app.window.allocated_height * 0.5
+    @list_widget.columns =  { {style: 'filename'} }
+    @command_line\add_widget 'completion_list', @list_widget
+
+  _auto_complete_file: (opts={}) =>
+    if not @list_widget
+      @_initialize_list_widget!
+
+    if not @list_widget.showing
+      @list_widget\show!
+
+    if opts.directories_only
+      @directory_reader = (dir) ->
+        dirs = [c for c in *dir.children when c.is_directory]
+        append dirs, 1, dir\join '.'
+        dirs
+    else
+      @directory_reader = (dir) -> dir.children
+
+    @_update_auto_complete!
+
+  _update_auto_complete: =>
+    return unless @list_widget and @list_widget.showing
+    last_part = last_text_part @command_line.text
+
+    if not last_part
+      @list_widget\hide!
+      return
+
+    path, unmatched = parse_path @directory.path .. File.separator .. last_part
+    @list_widget.matcher = file_matcher self.directory_reader(path), path
+    @list_widget\update unmatched
+    @list_widget_path = path
+    @list_widget_unmatched = unmatched
+
+  _auto_complete_command: (text) =>
+    if not @commands
+      @commands = available_commands!
+
+    if @list_widget
+      @list_widget\hide!
+
+    selected = interact.select
+      items: @commands
+      title: 'Commands'
+      columns: { { header: 'Commands', style: 'string' } }
+      submit_on_space: true
+      cancel_on_backspace: true
+      :text
+
+    if selected
+      @command_line\write selected.selection
+      @command_line\write ' '
+
+  _select_completion: =>
+    @list_widget\hide!
+    return unless @list_widget.selection
+
+    filename = @list_widget.selection.name
+    new_path = @list_widget_path / filename
+    @command_line.text = @command_line.text\sub(1, -#@list_widget_unmatched - 1)
+    @command_line\write filename
+    if not new_path.is_directory
+      @command_line\write ' '
+      @list_widget\hide!
+
+  _submit: =>
+    -- remove prompt so correct history is catpured
+    @command_line.prompt = ''
+    self.finish @directory.path, @command_line.text
+
+  keymap:
+    enter: =>
+      text = @command_line.text
+      cd_cmd = text\umatch '^%s*cd%s+'
+      if cd_cmd and @list_widget and @list_widget.showing
+        if (@list_widget.selection and @list_widget.selection.name == './') or not @list_widget.selection
+          dir = @command_line.text\umatch '^%s*cd%s+(.+)'
+          path = @directory / dir
+          if path.exists and path.is_directory
+            @_chdir path
+            @list_widget\hide!
+            @command_line.text = ''
+          return
+        @_select_completion!
+        return
+
+      if @list_widget and @list_widget.showing
```

Comment:
This almost duplicates the `if` above it. Perhaps get that out the way first at the start of the method?

```moonscript
unless @list_widget and @list_widget.showing
  @_submit!
  return
```